File: solutions/1239. Maximum Length of a Concatenated String with Unique Characters/1239.py
# Time complexity: O(2^n)
# Space complexity: O(n)
class Solution:
  def maxLength(self, arr: List[str]) -> int:
    def get_mask(s: str) -> int:
      mask = 0
      for c in s:
        bit = 1 << (ord(c) - ord('a'))
        if mask & bit:
          return 0
        mask |= bit
      
      return mask

    def dfs(masks: List[int], pos: int, used: int) -> int:
      max_len = bin(used).count('1')
      for i in range(pos, len(masks)):
        if masks[i] & used == 0:
          max_len = max(max_len, dfs(masks, i + 1, used | masks[i]))
      
      return max_len

    masks = [get_mask(s) for s in arr if get_mask(s) > 0]
    return dfs(masks, 0, 0)

File: solutions/1239. Maximum Length of a Concatenated String with Unique Characters/1239.cpp
// Time complexity: O(2^n)
// Space complexity: O(n)
class Solution {
 public:
  int maxLength(const vector<string>& arr) {
    vector<bitset<26>> masks;

    for (string_view s : arr) {
      const bitset<26> mask = getMask(s);
      if (mask.count() == s.length())
        masks.push_back(mask);
    }

    return dfs(masks, 0, bitset<26>());
  }
 private:
  int dfs(const vector<bitset<26>>& masks, const int s, 
          const bitset<26> used) {
    int res = used.count();
    for (int i = s; i < masks.size(); ++i)
      if (!(used & masks[i]).any())
        res = max(res, dfs(masks, i + 1, used | masks[i]));

    return res;        
  }

  bitset<26> getMask(string_view s) {
    bitset<26> mask;
    for (const char c : s)
      mask.set(c - 'a');

    return mask;    
  } 
};